FPUv1

Indanz opened the pull request:
First, simplest version of semi-lazy FPU switching.

TODO:
- Check Armv8 virtualisation assumptions.

- Try non-NULL ksActiveFPUState.
- Try global ksFPUEnabled.
- Try alternative domain switch variants.

(This pull request is for my local repo, not seL4! I'll create a proper pull request when all issues are resolved.)

Indanz committed:
Aarch64: Align FPU vregs

Aligning saves one cycle per LDP/STP instruction.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
FPU: ksCurFPUOwner instead of ksActiveFPUState

Trivial change, except for x86 virtualisation and bootup.

Normally the seL4_TCBFlag_fpuDisabled TCB flag is used to decide whether to
enable or disable the FPU. However, with x86 virtualisation there is only
one TCB. Use the task flag for the host and always enable FPU for the guest.

When either the host or the guest's FPU is loaded, ksCurFPUOwner will point
to our TCB. saveFpuState and loadFpuState check fpu_active and do the right
thing depending on the current FPU user. To make this work it was necessary
to pass arch_tcb_t to saveFpuState and loadFpuState.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
Add seL4_TCB_SetFlags Syscall

Add flags to tcb_t and the seL4_TCBFlag_fpuDisabled flag.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
FPU: Save and restore FPU state based on TCB flag

Remove fault-based FPU state saving and loading.

TODO: Check if armv7-a vcpu_enable() is still correct.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
Domain: Avoid touching cross-domain FPU state

Save current FPU state on domain switch.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz commented on include/arch/x86/arch/machine/fpu.h:
```diff
@@ -114,7 +114,7 @@ static inline void disableFpu(void)
 #ifdef CONFIG_VTX
 static inline bool_t vcpuThreadUsingFPU(tcb_t *thread)
 {
-    return thread->tcbArch.tcbVCPU && &thread->tcbArch.tcbVCPU->fpuState == NODE_STATE(ksActiveFPUState);
```
So this line is currently incorrect, and probably the whole reason why `ksActiveFPUState` wasn't a TCB to begin with.

Indanz commented on include/arch/x86/arch/machine/fpu.h:
```diff
@@ -114,7 +114,7 @@ static inline void disableFpu(void)
 #ifdef CONFIG_VTX
 static inline bool_t vcpuThreadUsingFPU(tcb_t *thread)
 {
-    return thread->tcbArch.tcbVCPU && &thread->tcbArch.tcbVCPU->fpuState == NODE_STATE(ksActiveFPUState);
+    return thread->tcbArch.tcbVCPU && thread == NODE_STATE(ksActiveFPUThread);
 }
```
So this line is currently incorrect, and probably the whole reason why `ksActiveFPUState` wasn't a TCB to begin with.

Indanz commented on include/arch/x86/arch/machine/fpu.h:
```diff
@@ -114,7 +114,7 @@ static inline void disableFpu(void)
 static inline bool_t vcpuThreadUsingFPU(tcb_t *thread)
 {
-    return thread->tcbArch.tcbVCPU && &thread->tcbArch.tcbVCPU->fpuState == NODE_STATE(ksActiveFPUState);
+    return thread->tcbArch.tcbVCPU && thread == NODE_STATE(ksActiveFPUThread);
```
So this line is currently incorrect, and probably the whole reason why `ksActiveFPUState` wasn't a TCB to begin with.

lsf37 replied:
Can you elaborate more why this line is currently incorrect? I'm probably missing too much background on how x86 virtualisation works.

Indanz replied:
It replaced a check of `thread->tcbArch.tcbVCPU->fpuState` with an (implicit) check of `thread->tcbArch.tcbContext.fpuState`.

On x86 the thread switches between VM monitor and VM guest mode, both having their own FPU state, which we shouldn't mix up by accident.

lsf37 commented on include/object/structures.h:
```diff
@@ -300,6 +299,11 @@ struct tcb {
     struct tcb *tcbEPNext;
     struct tcb *tcbEPPrev;
 
+#ifdef CONFIG_HAVE_FPU
+    /* Currently only used for seL4_TCBFlag_fpuDisabled */
+    seL4_TCBFlag flags;
+#endif
```
Should we keep the field in the TCB struct in all configs instead?

It's an additional field, but it doesn't change the power of 2 of the TCB size, so shouldn't affect anything else. There are two sides for that for verification (will need small changes in otherwise unaffected verification configs, but will make everything more uniform). Raf, Corey any opinion on that?

corlewis replied:
My first reaction is that making this uniform across configs would be nice.

I might be missing some context, but since I'm assuming that we're implementing it this way to be able to support adding other flags in the future, then it makes sense to me to do it properly now instead of needing to come back in the future if/when we add generic flags. As you said, it shouldn't add much verification work now and it might help us avoid accidentally using any arch assumptions in the proofs.

Xaphiosis replied:
I like uniformity. It indeed doesn't bump up TCB size, and I don't think there's much benefit to dropping it.

Indanz replied:
If size is a problem, we can make the bool variables one byte instead of a `word_t`.

Xaphiosis replied:
Given the next thing in the struct is very likely going to be `word_t`-aligned, e.g. scheduling context on MCS, a byte would need padding anyway, so might as well keep `word_t`. I think there's still plenty of space in the TCB.

Indanz replied:
Actually, we have the following 1 byte fields next to each other in `tcb_t`: `tcbDomain`, `tcbMCP`, `tcbPriority`. Making them one byte would save two words, even with padding. `tcbAffinity` could be 1, 2 or 4 bytes too, and that would save another word.

Saving 24 bytes is almost half a cache line, so it would be worth it just for the increased cache hit rate.

Also, we're quite close to the size limit on some architectures, e.g. I couldn't align `fpuState` to 64 bytes on Aarch64, only to 32 bytes.

(Real savings could be made if `cte_t` is allowed to be after `tcb_t` though, or if the unused area is replaced with explicit padding.)

Xaphiosis replied:
I think we'd be fine with squashing it among the 3 other 8-bit values, not so sure if this is the time to start rearranging the TCB for size savings. I'm surprised to hear we're getting close to the size limit though!

lsf37 reviewed (commented):
Overall this looks much more sane than the current way of doing things..

lsf37 commented:
Is there no way to keep the current FPU owner to type `tcb_t`? Pointers to the insides of structs are a major pain in the verification and removing this was one of the (many) motivations for changing things in the first place.

Indanz commented:
The problem is that x86 has two FPU states: One for the monitor, and one for the guest VM. If we make the current FPU owner a `tcb_t`, it can't point to the guest VM FPU state. (It looks like the vcpu FPU state is not properly cleaned up on vcpu deletion currently, `clearCurrentVCPU` doesn't clear FPU state.)

I tried fudging it with a `fpu_valid` bool in vcpu and setting the current FPU owner to the TCB as usual, but then on a task switch the guest FPU state will be saved to the wrong area, overwriting the VMM FPU state.

Only other alternatives I can think of is to either do eager FPU switching for VCPUs, but that may slow down all VM calls, or change x86 virtualisation and use two TCB's, one for the guest and one for the monitor, like Arm does.

I'm currently in the progress of removing lazy, fault based FPU switching for VM guests. To me it makes sense to always enable the FPU for VCPUs, both for x86 and Arm.

Indanz commented:
Anyway, I've the feeling I have to spend some time adding virtualisation tests to see if everything works correctly.

Indanz commented:
> Is there no way to keep the current FPU owner to type `tcb_t`? Pointers to the insides of structs are a major pain in the verification and removing this was one of the (many) motivations for changing things in the first place.

What we could do is changing `saveFpuState` and `loadFpuState` to take a `arch_tcb_t`, and then the x86 arch code can handle the x86 vcpu corner case. I'll give that a go and see if I run into any issues.

Indanz committed:
x86, HYP: Cleanup VCPU FPU state

Otherwise ksActiveFPUState may point to freed memory after a vcpu
with active FPU state gets deleted.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
x86, HYP: Remove Lazy VCPU FPU Switching

On x86 both the VM monitor as the VM guest have their own FPU state.
The FPU of the monitor is controlled with seL4_TCBFlag_fpuDisabled
TCB flag. For the guest running on the VCPU the FPU will always be
enabled.

Signed-off-by: Indan Zupancic <indan@nul.nu>

lsf37 commented:
> The problem is that x86 has two FPU states: One for the monitor, and one for the guest VM. If we make the current FPU owner a `tcb_t`, it can't point to the guest VM FPU state. 

Right, thanks for that.

> (It looks like the vcpu FPU state is not properly cleaned up on vcpu deletion currently, `clearCurrentVCPU` doesn't clear FPU state.)

Entirely possible, there has been no verification yet for x86 virtualisation, and really only the infoflow proofs would find this specific kind of problem.


> I tried fudging it with a `fpu_valid` bool in vcpu and setting the current FPU owner to the TCB as usual, but then on a task switch the guest FPU state will be saved to the wrong area, overwriting the VMM FPU state.
> 
> Only other alternatives I can think of is to either do eager FPU switching for VCPUs, but that may slow down all VM calls, or change x86 virtualisation and use two TCB's, one for the guest and one for the monitor, like Arm does.
> 
> I'm currently in the progress of removing lazy, fault based FPU switching for VM guests. To me it makes sense to always enable the FPU for VCPUs, both for x86 and Arm.

Current AArch32 hyp verification doesn't have FPU, so we do need a config where it is possible for VCPU threads to not have FPU (but configured globally in that case). I guess it may make sense for VCPU threads to also have FPU on when FPU is available in general since VCPU is virtualising way more context already anyway, but I haven't fully thought through that yet.

Indanz commented:
> Current AArch32 hyp verification doesn't have FPU, so we do need a config where it is possible for VCPU threads to not have FPU (but configured globally in that case).

Those will have `CONFIG_HAVE_FPU` undefined, so no new config option is necessary.

> I guess it may make sense for VCPU threads to also have FPU on when FPU is available in general since VCPU is virtualising way more context already anyway, but I haven't fully thought through that yet.

But on Arm we have a separate TCB, so can just use the same FPU code as for normal tasks, no need to do anything special I think. So what I said before really only applies to x86.

Indanz commented:
> Is there no way to keep the current FPU owner to type `tcb_t`? Pointers to the insides of structs are a major pain in the verification and removing this was one of the (many) motivations for changing things in the first place.

Okay, done. I can also go one step further and change `saveFpuState` and `loadFpuState` to take a `tcb_t` instead of a `arch_tcb_t` to avoid the pointers inside structs in `switchLocalFpuOwner`, if that makes verification easier.

corlewis commented on src/kernel/thread.c:
```diff
@@ -307,6 +307,10 @@ static void nextDomain(void)
 #ifdef CONFIG_KERNEL_MCS
     NODE_STATE(ksReprogram) = true;
+#endif
+#ifdef CONFIG_HAVE_FPU
+    /* Save FPU state now to avoid touching cross-domain state later: */
+    switchLocalFpuOwner(NULL);
 #endif
```
I've been working on something similar recently, which involved flushing the vcpu before switching domain. Based on that, it helps verification if we can keep `nextDomain`  so that it only modifies the domain related parts of the state, and separate out the other state updates that we're adding. 

How would you feel about something like a `prepareNextDomain` , in the same way that I've done in https://github.com/seL4/seL4/compare/master...corlewis:vcpu_flush.

Indanz replied:
That would be fine, lazy VCPU switching has the same issues as lazy FPU switching.

I know it's not verified, but for x86 you want to call `clearCurrentVCPU` on domain switch.

Do you want me to merge your commit with the vcpu flushing too or should I just add the same `Arch_prepareNextDomain` with FPU clearing?

corlewis replied:
Thanks, that's good to know for x86, I hadn't looked into other architectures at all yet.

Probably best not to merge it, I've updated the proofs for that change on aarch64 but there's a related issue that's causing problems and is going to take a while to work through. If you're ok with that structure then adding `Arch_prepareNextDomain` with the FPU clearing would be good, and then once the VCPU work is finished they can both be cleared in the same function.

Indanz replied:
I've added a `prepareNextDomain`, as it seems more config dependent than arch dependent. We can change it later if you want.

corlewis replied:
That's fine, I think I went with the arch naming because I came at it from the verification perspective, where we treat different configs like this very similarly to architectures.

Xaphiosis commented on include/model/statedata.h:
```diff
@@ -73,11 +73,10 @@ NODE_STATE_DECLARE(sched_context_t, *ksIdleSC);
-NODE_STATE_DECLARE(user_fpu_state_t *, ksActiveFPUState);
-/* Number of times we have restored a user context with an active FPU without switching it */
-NODE_STATE_DECLARE(word_t, ksFPURestoresSinceSwitch);
+/* Current thread using the FPU, or NULL if FPU state is invalid */
```
suggestion: `The thread currently using the FPU, ...` (I was momentarily confused that this indicated the current thread is using the FPU)

Indanz replied:
Good point, but any use of "current" may be confusing. What about `Last thread that used the FPU`?

Xaphiosis replied:
That phrase happens to be true, but is that what you want to communicate to people? There's an association/ownership aspect here that might be important.
Otherwise sure, that will work.

Indanz replied:
I'm trying to make it clearer what the variable means. I could use the more verbose:

`Thread whose current state is installed in the FPU, or NULL if the FPU is currently invalid`

Indanz replied:
I don't like that either, what about:

`The thread using the FPU, or NULL if FPU state is invalid`?

To the point and short, with no room for confusion.

corlewis replied:
As a sort of data point for this, in the specifications I'm currently calling the equivalent variable `arm_current_fpu_owner`. Do you think a name like that makes sense and could be used here, or should I be avoiding 'current' there as well?

Xaphiosis replied:
I like the idea of something like `currentFPUOwner` (note: the `lower_underscore` style is specific to our abstract spec)

Indanz replied:
For consistency with the naming of other current states, it should then be `ksCurFPUOwner`. I'm fine with changing it, I'm not attached to the `ksActiveFPUThread` naming.

Indanz committed:
x86, FPU: Fix XSAVES

Add config choice.

Get rid of dubious FPU state headers, we don't need them:
- XCOMP_BV_COMPACTED_FORMAT is set by xsavec or xsaves.
- We can init MXCSR with the vldmxcsr instruction.

All memory should be zeroed already, no need to do it again.

See also issue #179.

Signed-off-by: Indan Zupancic <indan@nul.nu>

corlewis commented on src/kernel/thread.c:
```diff
@@ -299,6 +300,14 @@ void doNBRecvFailedTransfer(tcb_t *thread)
     setRegister(thread, badgeRegister, 0);
 }
 
+static void prepareNextDomain(void)
+{
+    if (config_set(CONFIG_HAVE_FPU)) {
+        /* Save FPU state now to avoid touching cross-domain state later */
+        switchLocalFpuOwner(NULL);
```
After some thought, I've convinced myself that the related issue that I've been dealing with for VCPUs will be a problem for FPUs as well. Basically, when setting the domain of a tcb we need to be careful and sometimes save the state there first as well. Otherwise, we might have changed the domain of the thread that owns the current FPU state, and when we then save it later on we would be writing to cross-domain state.

This is sort of an edge case, as any realistic systems where we care about controlling information flow shouldn't be moving threads between domains like that, but it's something that we need to handle to make it possible to complete the proofs.

A first version of the equivalent change needed for VCPUs can be seen at https://github.com/seL4/seL4/commit/c057459b2d46482f04d676a6b15b98c2a5fedb51, although please keep in mind that I haven't updated the proofs for that yet, or even checked that it compiles, so that code might change.

Indanz replied:
Perhaps that's necessary because of limitations of the proof, but conceptually I disagree it's needed:
- All FPU and VCPU state is the current domain's state.
- Changing domain of a task stops it from running and hence stops it from changing that state.
- That state will be written in the current domain when changing domains.
- Until the task runs again, that state is effectively usable by both the current domain and the new domain.

corlewis replied:
I actually agree with you that conceptually this shouldn't be needed, but that last point doesn't really work for the proofs, we currently aren't able to handle state like this that could be used by separate domains. I'll have one more think to see if there's anything clever we can do to avoid needing to make this change but I'm not confident.

corlewis commented:
As far as I can tell, it looks like you've removed all uses of `isFpuEnable` (and the `isFPUEnabledCached` variable that it refers to). Does that sound right to you? If it does, do you have any guesses about whether it's something that we want to keep around for possible future uses?

If you're curious, the source of these questions is that we're trying to work out how much detail we want to add to the specifications for this and it's unclear whether we need an equivalent for `isFpuEnable`. Regardless of what we end up deciding, the comment before `isFpuEnable` about the model not including lazy FPU switching and it not being translated can be removed now.

corlewis commented on include/machine/fpu.h:
```diff
@@ -16,43 +16,26 @@
+    return thread == NODE_STATE_ON_CORE(ksActiveFPUThread, thread->tcbAffinity);
 }
 
+/* Called without global lock held! */
 static inline void FORCE_INLINE lazyFPURestore(tcb_t *thread)
 {
```
Unfortunately, now that we're actually modelling lazy FPU switching, this being called from `restore_user_context` and `fastpath_restore` is going to cause major difficulties. The way those functions are modelled in the specifications is very unique, and won't be able to handle `lazyFPURestore` with its conditionals and reading and writing to kernel state.

Would it be possible to call this in the scheduler instead, as part of `switchToThread`, in roughly the same way that VCPU state being switched is handled? That would be much better from the verification perspective.

Xaphiosis replied:
To also add to this: restoring thread user context used to be unconditional and did not modify kernel state, which is why in verification we could sweep it under the rug as part of the kernel enter-exit procedure. As Corey says, `lazyFPURestore` doesn't meet the criteria for safe wrap-and-forget, so `switchToThread` looks like a good place (directly before or after `Arch_switchToThread`).

Indanz replied:
I suspected this would be a problem. I'll have a look at moving it.

Xaphiosis commented:
> As far as I can tell, it looks like you've removed all uses of `isFpuEnable` (and the `isFPUEnabledCached` variable that it refers to). Does that sound right to you? If it does, do you have any guesses about whether it's something that we want to keep around for possible future uses?
> 
> If you're curious, the source of these questions is we're trying to work out how much detail we want to add to the specifications for this and it's unclear whether we need an equivalent for `isFpuEnable`. Regardless of what we end up deciding, the comment before `isFpuEnable` about the model not including lazy FPU switching and it not being translated can be removed now.

To add to this: from a verification perspective it makes a difference whether we make impactful control flow decisions based on some aspect of the "machine state". If we do, then weneed to model  that state (in this case, whether FPU is enabled), which was what happened previously with `isFpuEnable` and the cache variable.

Indanz commented:
> As far as I can tell, it looks like you've removed all uses of `isFpuEnable` (and the `isFPUEnabledCached` variable that it refers to). Does that sound right to you? If it does, do you have any guesses about whether it's something that we want to keep around for possible future uses?

Yes, I was aware of that. The only user left is riscv, but arguably it can also use its arch specific `isFPUEnabledCached` instead, so it doesn't need to be part of the generic FPU code.

Currently I unconditionally enable or disable the FPU, even if it's already enabled or disabled. This is what the old code also did. If we want to avoid unnecessary FPU enable/disables, we need to add a global FPU enabled flag, which would replace`isFpuEnable()` as it's currently used.

So I see a use for the global FPU enabled flag, as it may also help in proving the code does the right thing, as without it there is no concept of the FPU being enabled or not currently (I imagine you want to proof e.g. that `saveFpuState` and `loadFpuState` are only called with FPU enabled).

> If you're curious, the source of these questions is that we're trying to work out how much detail we want to add to the specifications for this and it's unclear whether we need an equivalent for `isFpuEnable`. Regardless of what we end up deciding, the comment before `isFpuEnable` about the model not including lazy FPU switching and it not being translated can be removed now.

Currently the HW FPU model is:

- enableFpu()
- disableFpu()
- saveFpuState()
- loadFpuState()

(And I may add a switchFpuState() which equals save + load.)

What I've seen with x86 and Arm32 virtualisation is that it may be helpful if the arch code knows what the final decision is. Now e.g. enableFpu() can be part of switchLocalFpuOwner() or the only thing that happens in lazyFPURestore(). For that making `switchLocalFpuOwner` arch specific would probably be enough.

In the end it depends on how much you want to verify and what you want to proof.

corlewis commented:
> So I see a use for the global FPU enabled flag, as it may also help in proving the code does the right thing, as without it there is no concept of the FPU being enabled or not currently (I imagine you want to proof e.g. that `saveFpuState` and `loadFpuState` are only called with FPU enabled).

Ah, I hadn't made the connection before that `saveFpuState` and `loadFpuState` need the FPU to be enabled to be called. That makes complete sense, and also explains why `switchLocalFpuOwner` first enables the FPU.

You're right that we'll want to prove that. I might add assertions to our models of `saveFpuState` and `loadFpuState` to force us to do so.

corlewis commented on include/arch/x86/arch/machine/fpu.h:
```diff
@@ -59,8 +34,17 @@ static inline uint32_t xsave_features_low(void)
 {
+    user_fpu_state_t *dest = &arch_tcb->tcbContext.fpuState;
+#ifdef CONFIG_VTX
+    struct vcpu *vcpu = arch_tcb->tcbVCPU;
+
+    if (vcpu->fpu_active) {
```
I might be missing some x86 details that means that the threads being passed in here and in `loadFpuState` must have a vcpu, but don't you need to check that the vcpu exists first?
```suggestion
    struct vcpu *vcpu = arch_tcb->tcbVCPU;

    if (vcpu && vcpu->fpu_active) {
```

Indanz replied:
Indeed, not sure how I missed that.

corlewis commented:
> > Is there no way to keep the current FPU owner to type `tcb_t`? Pointers to the insides of structs are a major pain in the verification and removing this was one of the (many) motivations for changing things in the first place.
> 
> Okay, done. I can also go one step further and change `saveFpuState` and `loadFpuState` to take a `tcb_t` instead of a `arch_tcb_t` to avoid the pointers inside structs in `switchLocalFpuOwner`, if that makes verification easier.

Just coming back to this, if you're still ok with it then changing them to use a `tcb_t` like that would almost definitely help us.

Indanz commented:
> Ah, I hadn't made the connection before that `saveFpuState` and `loadFpuState` need the FPU to be enabled to be called. That makes complete sense, and also explains why `switchLocalFpuOwner` first enables the FPU.
> 
> You're right that we'll want to prove that. I might add assertions to our models of `saveFpuState` and `loadFpuState` to force us to do so.

Well, outside of arch machine and boot code, `switchLocalFpuOwner` is the only one enabling or disabling the FPU, so proving it has limited value. But I guess you might want to proof that, if the current task is allowed to use the FPU, the FPU is actually enabled. Would having explicit state for that help with the proofs or not really?

> Just coming back to this, if you're still ok with it then changing them to use a tcb_t like that would almost definitely help us.

I don't know how the proofs work, but from a C point of view I have, passing `arch_tcb_t` implies those functions only modify arch state, while passing `tcb_t` implies it might change anything in `tcb_t`, so that's why I passed in `arch_tcb_t`. Happy to change it to `tcb_t` though.

Indanz committed:
FPU: Pass tcb_t to saveFpuState and loadFpuState

To ease verification.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
domain: Save FPU state when changing domain

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
FPU: Remove Unused isFpuEnable

Unused on ARM, on RISC-V it is actually the final decision and not
a cache.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
FPU: Do lazyFPURestore in switchToThread

To ease verification.

Fairly trivial change, exept for x86 of course.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
FPU: Rename fpuThreadDelete to fpuRelease

It's not only used when deleting a thread any more.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
x86, FPU: Handle obscure corner cases

Also release FPU when dissociating a vcpu, otherwise ksCurFPUOwner
may be wrong.

fpuRelease() sets ksCurFPUOwner to NULL, if this happens before
we return to the host, it may end up with the FPU disabled and no
FPU state loaded. Instead of hunting down and handling all obscure
corner cases where this might happen (dissociating vcpus, cross-core
operations), just check for this in vcpu_fpu_to_host().

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
RISC-V, FPU: Only save FPU state if FS_DIRTY

With trap based FPU switching the dirty bit would always be set
because the trap is caused by an FPU instruction, which would
enable the FPU and let the task change its current FPU state.

Now with task flag based FPU switching the dirty bit will only
be set when a task actually used the FPU, so checking it in
saveFpuState() is an easy optimisation.

A possible optimisation for loadFpuState() is to check the INITIAL
bit, then FPU memory loads can be replaced with a loop of setting
all FP regs to zero. However, FPU Load-Immediate Instructions are
part of the "Zfa" Extension and without them there isn't much gain,
so don't bother with the INITIAL optimisation.

Add comments to set_fs_clean() calls to clarify that they're called
to enable the FPU, not to change the state to CLEAN, which is done
by indirectly set_tcb_fs_state().

Signed-off-by: Indan Zupancic <indan@nul.nu>

corlewis commented on src/kernel/thread.c:
```diff
@@ -299,6 +300,22 @@ void doNBRecvFailedTransfer(tcb_t *thread)
 
+static void prepareSetDomain(tcb_t *tptr, dom_t dom)
+{
+    if (config_set(CONFIG_HAVE_FPU)) {
+        /* Save FPU state now to avoid touching cross-domain state later */
+        fpuRelease(tptr);
+    }
```
I'll keep on trying to find an alternative way of handling this but for now we do need it for verification, so thank you for adding it. However, I think it would be better with a check to avoid saving the fpu state if for whatever reason we're setting the domain of the thread to the current domain. That would help us avoid having to think through and reason about whether there's any code paths that lead to resuming a thread with its fpu state unexpectedly cleared.
```suggestion
    if (config_set(CONFIG_HAVE_FPU)) {
        if (ksCurDomain != dom) {
            /* Save FPU state now to avoid touching cross-domain state later */
            fpuRelease(tptr);
        }
    }
```

Indanz replied:
Good find! Seems that that case is currently handled, but perhaps not intentionally:

setdomain calls rescheduleRequired, which set `ksSchedulerAction = SchedulerAction_ChooseNewThread`.
After handling a syscall, the call chain is schedule -> scheduleChooseNewThread ->  chooseThread which calls either switchToThread or switchToIdleThread.

However, the other call to switchToThread in schedule has `assert(candidate != NODE_STATE(ksCurThread))`, which to me hints it's an unexpected case. Then again, the assert isn't in switchToThread itself.

Anyway, wouldn't it be better to do the `ksCurDomain != dom` check in decodeDomainInvocation and skip calling setdomain altogether?

Indanz replied:
Or an `if (tptr->tcbDomain == dom) { return; }` in setDomain itself.

corlewis replied:
Oh, good thinking! The second approach of doing the check and returning in `setDomain` would be better for verification, or at least require less other changes. There are some technicalities with our specifications that I'm a bit concerned about though, so it might be best to leave this for after the current deadline for when I have time to explore it in more detail.

Indanz committed:
Update src/kernel/thread.c

Co-authored-by: Corey Lewis <corlewis@gmail.com>

Indanz commented:
Created a [PR](https://github.com/seL4/seL4/pull/1325) on the seL4 repo, closing this one. Please continue any discussions over there, thanks!
